rllib/agents/dqn/apex.py

```diff
@@ -283,6 +374,256 @@ def add_apex_metrics(result: dict) -> dict:
+                remote_kwargs=[{"replay_actors": self.replay_actors}]
+                * len(self.workers.remote_workers()),
+            )
+        return num_samples_ready_dict
+
+    def update_workers(self, _num_samples_ready: Dict[ActorHandle, int]) -> int:
+        """Update the remote workers that have samples ready.
+
+        Args:
+            _num_samples_ready: A mapping from ActorHandle (RolloutWorker) to
+                the number of samples returned by the remote worker.
+        Returns:
+            The number of remote workers whose weights were updated.
+        """
+        max_steps_weight_sync_delay = self.config["optimizer"]["max_weight_sync_delay"]
+        # Update our local copy of the weights if the learner thread has updated
+        # the learner worker's weights
+        if self.learner_thread.weights_updated:
+            self.learner_thread.weights_updated = False
+            weights = self.workers.local_worker().get_weights()
+            self.curr_learner_weights = ray.put(weights)
+        with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
+            for (
+                remote_sampler_worker,
+                num_samples_collected,
+            ) in _num_samples_ready.items():
+                self.steps_since_update[remote_sampler_worker] += num_samples_collected
+                if (
+                    self.steps_since_update[remote_sampler_worker]
+                    >= max_steps_weight_sync_delay
+                ):
+                    remote_sampler_worker.set_weights.remote(
+                        self.curr_learner_weights,
+                        {"timestep": self._counters[STEPS_TRAINED_COUNTER]},
+                    )
+                    self.steps_since_update[remote_sampler_worker] = 0
+                self._counters["num_weight_syncs"] += 1
+
+    def sample_from_replay_buffer_place_on_learner_queue_non_blocking(
+        self, num_samples_collected: Dict[ActorHandle, int]
+    ) -> None:
+        """Get samples from the replay buffer and place them on the learner queue.
+
+        Args:
+            num_samples_collected: A mapping from ActorHandle (RolloutWorker) to
+                number of samples returned by the remote worker. This is used to
+                implement training intensity which is the concept of triggering a
+                certain amount of training based on the number of samples that have
+                been collected since the last time that training was triggered.
+
+        """
+        num_samples_collected = sum(num_samples_collected.values())
+        self.curr_num_samples_collected += num_samples_collected
+        if self.curr_num_samples_collected > self.config["train_batch_size"]:
+            self.curr_num_samples_collected = 0
+            training_intensity = int(self.config["training_intensity"] or 1)
+            # for _ in range(training_intensity):
+            # rand_actor = random.choice(self.replay_actors)
```

Comment:
debugging code?

Reply:
ah not exactly, that is the code that should actually be in there. Basically right now I sample from all the replay buffers and place their results on the learner queue, but this is a bad design because it means that the amount of samples placed on the learner queue is a function of the number of replay buffers that there are. This is how we do this right now also, but instead, I should just be sampling from a random replay buffer actor every time i need replay buffer samples.